File: web/src/lib/redondeo.ts
// Redondeo a la unidad (peso) en multiplicaciones monetarias. Gemelo de
// apu_tool/nucleo/redondeo.py. Regla: cada producto monetario se redondea a la
// unidad más cercana (medio hacia arriba); un producto positivo que redondearía
// a 0 se fija en 1 (nada en $0 por redondeo); un 0 genuino queda en 0.
export function mulRedondeado(a: number, b: number): number {
  const p = a * b;
  // `!(p > 0)` y NO `p <= 0`: toda comparación con NaN es false, así que un NaN se
  // colaría a Math.round(NaN) = NaN y se pintaría "NaN" en la tabla. Mismo arreglo
  // que en el gemelo de Python (apu_tool/nucleo/redondeo.py).
  if (!(p > 0)) return 0;
  // Math.round en JS es medio-hacia-+∞: Math.round(0.5)=1, Math.round(1312.5)=1313.
  const r = Math.round(p);
  return r !== 0 ? r : 1;
}

File: apu_tool/nucleo/redondeo.py
"""Redondeo a la unidad (peso) en multiplicaciones monetarias.

Regla del negocio: no se trabaja con decimales en dinero. El resultado de CADA
multiplicación monetaria se redondea a la unidad más cercana (medio hacia arriba).
Si el producto es positivo pero redondearía a 0, se fija en 1 (invariante "nada en
$0": el redondeo nunca hace desaparecer un costo real). Un 0 genuino queda en 0.

Módulo puro (solo stdlib). No toca dinero hacia la IA (invariante #1 no aplica aquí).
"""
from __future__ import annotations

import math


def mul_redondeado(a: float, b: float) -> int:
    """Multiplica y redondea a la unidad más cercana, medio hacia arriba.

    - ``a * b <= 0``  -> 0 (0 genuino; lo marca la alerta de costeo).
    - ``a * b > 0`` que redondearía a 0 -> 1 (nada en $0 por redondeo).
    - resto -> ``floor(a * b + 0.5)``.
    """
    p = a * b
    if p <= 0:
        return 0
    r = math.floor(p + 0.5)
    return r if r != 0 else 1
